pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+	// Controllers to locate the previous representation  of the endpoint and check
+	// for changes to its metadata that could result in changes to the subsetEndpoints.
+	allEndpoints map[string]*Endpoint
+
+	// subsetDefinitions holds the metadata, basically the set of lable name values
+	// that ought to match corresponding attributes of the endpoint.
+	// The actual subset of allEndpoints that match these subset definitions is
+	// stored in subsetEndpoints. Mesh uses this map for identifying config changes
+	// to the subset's metadata. The key of this map is the Subset name.
+	subsetDefinitions map[string]*route.Subset
+
+	// subsetEndpoints allows for look-ups from the Subset name to a set of Endpoint UIDs.
+	// The UIDs in the set are guaranteed to exist in allEndpoints and the
+	// corresponding LbEndpointis guaranteed to satisfy the labels in the corresponding
+	// Subset.
+	subsetEndpoints subsetEndpoints
+
+	// reverseAttrMap provides reverse lookup from label name > label value >
+	// endpointUIDSet. All Endpoints in the set are guaranteed to match the
+	// attribute with the label name and value. Mesh uses the map for quickly
+	// updating subsetEndpoints in response to subset label changes.
+	reverseAttrMap attributeValues
+
+	// reverseEpSubsets provides a reverse lookup from endpoint key to the set of
+	// subsets this endpoint matched to. All subset keys associated with the endpoint
+	// are guaranteed to exist in subsetDefinitions. Mesh uses this to reverse
+	// lookup all subsets impacted by changes to Endpoint metadata.
+	reverseEpSubsets endpointSubsets
+
+	// Mutex guards guarantees consistency of updates to members shared across
+	// threads.
+	mu sync.RWMutex
+}
+
+// Endpoint is a platform independent representation of a Mesh Endpoints that
+// uses Envoy v2 API's LbEndpoint as its internal implementation. It also
+// provides utility methods intended for platform specific Controllers.
+type Endpoint xdsapi.LbEndpoint
+
+// endpointUIDSet is a unique set of Endpoint UIDs.
+type endpointUIDSet map[string]bool
+
+// valueEndpointSet associates a label value with an endpointUIDSet.
+type valueEndpointSet map[string]endpointUIDSet
+
+// attributeValues is a reverse lookup map from label name > label value > endpointUIDs
+// that all have the same attribute matching the label name and label value.
+type attributeValues map[string]valueEndpointSet
+
+// subsetNameSet is a unique set of subset names.
+type subsetNameSet map[string]bool
+
+// endpointSubsets maps an endpoint UID to a set of matching subsets.
+type endpointSubsets map[string]subsetNameSet
+
+// subsetEndpoints maps an subset name to a set of matching endpoint UIDs.
+type subsetEndpoints map[string]endpointUIDSet
+
+// SubsetEvent is a pilot configuration event for notifying the specified
+// Subset is to be added, updated or deleted from Pilot depending on the
+// specified Event.
+type SubsetEvent struct {
+	Subset *route.Subset
+	Event  Event
```

Comment:
Somehow I can't find Event definition, but again it's a very big word... Change ? 
Would it be more efficient if we passed 3 subsets ( added  / updated / deleted ) instead ?

Reply:
I'm not sure we even need to process events one-by-one. If I add 1000 clusters, I can just list all 1000, not apply 1000 deltas. Where's this requirement coming from?

Reply:
No requirements - just advice from @mandarjog to not expect config changes in bits rather than in one big chunk. @kyessenov there are many ways to do this - one is we can have a map[Change][ ] *Subset or what I is there currently - which is the de-normalized  map. Both are intended for passing multiple subsets for a given call (versus one subset per call ... which would end up being wasteful) ... Which would you prefer - map[Event] [ ] *Subset or the de-normalized map in the form of an array.

Reply:
As I mentioned, I personally prefer 3 slices or maps - added, removed, modified, without the Event indirection.